_delphi_utils_python/delphi_utils/logger.py

```diff
@@ -84,10 +120,108 @@ def get_structured_logger(name=__name__,
     system_logger = logging.getLogger(name)
     if filename and not system_logger.handlers:
         system_logger.addHandler(logging.FileHandler(filename))
-    system_logger.setLevel(logging.INFO)
+    system_logger.setLevel(log_level)
     logger = structlog.wrap_logger(system_logger)
 
     if log_exceptions:
         handle_exceptions(logger)
 
     return logger
+
+
+
+
+class LoggerThread():
+    """
+    the bare structlog loggers are thread-safe but not multiprocessing-safe.
+    a `LoggerThread` will spawn a thread that listens to a mp.Queue
+    and logs messages from it with the provided logger,
+    so other processes can send logging messages to it
+    via the logger-like `SubLogger` interface.
+    the SubLogger even logs the pid of the caller.
+
+    this is good to use with a set of jobs that are part of a mp.Pool,
+    but isnt recommended for general use
+    because of overhead from threading and multiprocessing,
+    and because it might introduce lag to log messages.
+    """
+
+    # TODO: add checks to prevent use of a stopped thread?
+    # TODO: reduce level of a bunch of debugging logs (search "self.logger.info")
+
+    class SubLogger():
+        """
+        multiprocessing-safe logger-like interface
+        to convey log messages to a listening LoggerThread
+        """
+        def __init__(self, queue):
+            self.queue = queue
+        def _log(self, level, *args, **kwargs):
+            kwargs_plus = {'sub_pid': multiprocessing.current_process().pid}
+            kwargs_plus.update(kwargs)
+            self.queue.put([level, args, kwargs_plus])
+        # TODO: add log levels beyond `info`
+        def info(self, *args, **kwargs):
+            """log an INFO level message"""
+            self._log(logging.INFO, *args, **kwargs)
+
+
+    def get_sublogger(self):
+        """accessor method to retrieve a SubLogger for this LoggerThread"""
+        return self.sublogger
+
+    def __init__(self, logger, q=None):
+        self.logger = logger
+        if q:
+            self.msg_queue = q
+        else:
+            self.msg_queue = multiprocessing.Queue()
+
+        def logger_thread_worker():
+            self.logger.info('thread started')
```

Comment:
These and other thread-related messages seem like they should be `debug`-level or otherwise configurable, maybe through a boolean argument in `pool_and_threadedlogger`? For the Quidel indicator it would make sense to have them turned off by default.

Reply:
already taken care of!